src/fs/metadata_log_bootstrap.hh

```diff
@@ -0,0 +1,488 @@
+        }
+
+        // Only files may be linked as not to create cycles
+        if (not _metadata_log._inodes[entry.entry_inode].is_file()) {
+            return invalid_entry_exception();
+        }
+
+        if (not _metadata_log._inodes[entry.dir_inode].is_directory()) {
+            return invalid_entry_exception();
+        }
+        auto& dir = _metadata_log._inodes[entry.dir_inode].get_directory();
+
+        if (dir.entries.count(dir_entry_name) != 0) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_add_dir_entry(dir, entry.entry_inode, std::move(dir_entry_name));
+        // TODO: Maybe mtime_ns for modifying directory?
+        return now();
+    }
+
+    future<> bootstrap_create_inode_as_dir_entry() {
+        ondisk_create_inode_as_dir_entry_header entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.dir_inode) or inode_exists(entry.entry_inode.inode)) {
+            return invalid_entry_exception();
+        }
+
+        sstring dir_entry_name;
+        if (not _curr_checkpoint.read_sstring(dir_entry_name, entry.entry_name_length)) {
+            return invalid_entry_exception();
+        }
+
+        if (not _metadata_log._inodes[entry.dir_inode].is_directory()) {
+            return invalid_entry_exception();
+        }
+        auto& dir = _metadata_log._inodes[entry.dir_inode].get_directory();
+
+        if (dir.entries.count(dir_entry_name) != 0) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_create_inode(entry.entry_inode.inode, entry.entry_inode.is_directory, ondisk_metadata_to_metadata(entry.entry_inode.metadata));
+        _metadata_log.memory_only_add_dir_entry(dir, entry.entry_inode.inode, std::move(dir_entry_name));
+        // TODO: Maybe mtime_ns for modifying directory?
+        return now();
+    }
+
+    future<> bootstrap_delete_dir_entry() {
+        ondisk_delete_dir_entry_header entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.dir_inode)) {
+            return invalid_entry_exception();
+        }
+
+        sstring dir_entry_name;
+        if (not _curr_checkpoint.read_sstring(dir_entry_name, entry.entry_name_length)) {
+            return invalid_entry_exception();
+        }
+
+        if (not _metadata_log._inodes[entry.dir_inode].is_directory()) {
+            return invalid_entry_exception();
+        }
+        auto& dir = _metadata_log._inodes[entry.dir_inode].get_directory();
+
+        auto it = dir.entries.find(dir_entry_name);
+        if (it == dir.entries.end()) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_delete_dir_entry(dir, std::move(dir_entry_name));
+        // TODO: Maybe mtime_ns for modifying directory?
+        return now();
+    }
+
+    future<> bootstrap_rename_dir_entry() {
+        // TODO: implement it
+        assert(false && "Not implemented");
```

Comment:
About all these asserts - these are reachable for clients, right? Let's return an exception future instead, so that the process doesn't crash when calling a function - it will make testing easier.

Reply:
For now we don't intend to use / test this functions before implementing them, so I though that assert is fine.

Reply:
ok